Pass filter dropdowns as props

haideriqbal opened the pull request:
I've modified the component to pass dropdown filters as a prop.

haideriqbal committed:
- Add new dropdown for Projects
- Add new test for project dropdown filtering

haideriqbal committed:
- Fix styling issues

haideriqbal committed:
- Change project names to human readable format

haideriqbal committed:
- Pass filter dropdowns as props
- Update tests

lingyun1010 reviewed (changes requested):
I think we may not make clear about this story. The *props* should be passed via `html` file as the table header passed, so we can easily make changes, i.e. add a search box with dynamic title and filter strategy. For example, the [pivotal story](https://www.pivotaltracker.com/story/show/168868739) will be simply add a field in the html without modifying the js codes.

haideriqbal committed:
- Pass filter dropdowns as props from jsp
- Update tests

haideriqbal committed:
- Fix failing test

lingyun1010 commented:
It is improved but I think we still need to clarify the requirement further.
1. The title of the dropbox could be more user-friendly, which means we need to provide another field in the html/jsp props.
2. The options of the dropbox would be better retrieved from data (by filter/unique) rather than hard-coded.
You may like to refer to the table header props in the html file, such as `title` and `dataParam`. In this case, we do not require changing the json responses of `aaData`, where you add fields in [web-core](https://github.com/ebi-gene-expression-group/atlas-web-core/pull/24/files). 
We could benefit from defining the dropbox like the table headers, as it makes the component more general. If we have more dropboxes in future implementation, we don't have to modify backend or frontend at all, but just jsp.
Please feel free to come by if anything is not clear.

haideriqbal committed:
- Remove hard coded bits
- Adjust tests to new logic

lingyun1010 commented on src/ExperimentTable.js:
```diff
@@ -68,11 +77,27 @@ class ExperimentTable extends React.Component {
+        currentDropdownFilters[index].value = e.target.value : currentDropdownFilters.splice(index,1)
+      this.setState({
+        selectedDropdownFilters: currentDropdownFilters,
+        currentPage: 1
```
We may want to clear the table header search states once the filter dropdown has updated. For example, we set `species` as `homo sapiens` by default, but may result in conflict search results with `Protists` Kingdom.

haideriqbal replied:
Good point! I forgot about it! Fixed it now.

lingyun1010 commented on src/ExperimentTable.js:
```diff
@@ -68,11 +77,27 @@ class ExperimentTable extends React.Component {
+    const index = this.state.selectedDropdownFilters.findIndex(dropdown => dropdown.label === label)
+    if(index !== -1) {
+      selectedDropdown.value !== `` ?
+        currentDropdownFilters[index].value = e.target.value : currentDropdownFilters.splice(index,1)
```
Code style issue, we usually start a new line after `:`.

haideriqbal replied:
Fixed them

lingyun1010 commented on src/ExperimentTable.js:
```diff
@@ -68,11 +77,27 @@ class ExperimentTable extends React.Component {
+      label: label,
+      value: e.target.value
+    }
+    const currentDropdownFilters = this.state.selectedDropdownFilters
```
To copy a state to a `const`, we usually use the same variable name to keep it constant, and we use the format like  ```const {selectedDropdownFilters} = this.state``` to simplify codes.

haideriqbal replied:
Yes, it makes code cleaner too. Fixed it.

lingyun1010 commented on src/ExperimentTable.js:
```diff
@@ -68,11 +77,27 @@ class ExperimentTable extends React.Component {
+      currentDropdownFilters.push(selectedDropdown)
+      this.setState({
+        selectedDropdownFilters: currentDropdownFilters,
+        currentPage: 1
```
Same comments with the one on Line 92.

lingyun1010 commented on src/ExperimentTable.js:
```diff
@@ -103,33 +129,66 @@ class ExperimentTable extends React.Component {
+                  !obj.options.includes((data[key])) ? obj.options.push(data[key]) : true : true
+              let index = experimentTableFilters.findIndex(label => label === dropdown.dataParam)
+              index !== -1 ?
+                experimentTableFilters[index].options = obj.options : experimentTableFilters.push(obj)
```
This part of the codes is not readable. What do you want to generate for `experimentTableFileters`? Why do you need an `obj`? Why do you condition `obj.options.includes()` if you `_.uniq` those fields later? Could you think a better way to implement the logic?

haideriqbal replied:
I've improved the logic a bit and also renamed the variables to make it easier to understand.

lingyun1010 commented on src/ExperimentTable.js:
```diff
@@ -103,33 +129,66 @@ class ExperimentTable extends React.Component {
+        .some(value => value && value.toString().toLowerCase().includes(selectedSearch)))
+
+    const dropdownFilteredExperiments = selectedDropdownFilters.length === 0 ?
+      this.filter(this.sort(aaData), tableHeader) :
```
This line is reused and we could define a more sensible variable name for it, as it is not `dropdownFilteredExperiments`, which may mislead other developers.

haideriqbal replied:
I didn't understand this, `dropdownFilteredExperiments` does contain experiments filtered by selected dropdowns so it makes sense to name it accordingly.

lingyun1010 commented on src/TableSearchHeader.js:
```diff
@@ -1,22 +1,25 @@
-
+    {
+      dropdownFilters.map((dropdownFilter) => {
+        return (<div key={dropdownFilter.label} className={`small-12 medium-4 large-2 columns`}>
```
This `map` function could be reformated as `dropdownFilter => <div>...</div>`.

haideriqbal replied:
Modified it.

lingyun1010 reviewed (changes requested):
Please have a look into comments. We are almost there but need some further modifications. Thanks.

haideriqbal committed:
- Improve filtering logic
- Remove search query state on dropdown change
- Implemented other minor PR comments

haideriqbal committed:
- Refactor reading variable from state

haideriqbal committed:
- Assign table header filtered experiments to a variable

lingyun1010 commented on src/ExperimentTable.js:
```diff
@@ -137,37 +139,30 @@ class ExperimentTable extends React.Component {
+        label: ``,
+        options: []
+      }
+      dropdownFilter.label = filter.label
```
Why not set `label` when you initialize the `dropdownFilter`?

haideriqbal replied:
Agreed! it would be cleaner.

lingyun1010 commented on src/ExperimentTable.js:
```diff
@@ -137,29 +139,21 @@ class ExperimentTable extends React.Component {
+      let index = experimentTableFilters.findIndex(label => label === filter.dataParam)
+      index !== -1 ? experimentTableFilters[index].options = dropdownFilter.options :
+        experimentTableFilters.push(dropdownFilter)
+    })
```
It is improved a bit, but still very confusing. Would you like to reply the questions I commented before of put comments for the logic>

haideriqbal replied:
I want to create a structure for dropdowns like this:
`experimentTableFilters: [
{ label: ... ,
options: [...]
},
{...},
...
]`

I'm looping over the array of dropdowns being passed from props and creating an object of the structure `dropdownFilter: {
label: ...,
options: [...]
}` and then adding it in the array `experimentTableFilters` at the end of each loop.

To keep dropdown options distinct i'm checking if the option has already been added to the list `dropdownFilter.options`.

In the same way, I want to create one object per dropdown so before adding an `dropdownFilter` to `experimentTableFilters` I check if the `dropdownFilter` with same `label` exists.

lingyun1010 replied:
Thanks for the comments. Actually, I figured out a solution using `_.uniq` and `_.flatMap` to achieve what you want in two lines, but I will keep it secret for now. Would you like to have a go and we can discuss later?

haideriqbal committed:
- Improve logic of building experiment table filters

lingyun1010 commented:
The implementation seems good now. Could you please add some tests to increase the test coverage?

haideriqbal committed:
- Update test to increase coverage

alfonsomunozpomer commented on src/ExperimentTable.js:
```diff
@@ -103,33 +131,50 @@ class ExperimentTable extends React.Component {
+      return {
+        label: filter.label,
+        options: _.uniq(
+          _.flatMap(aaData.map(data => data[filter.dataParam])))
```
You can simplify `_.flatMap(aaData.map(...)` to `aaData.flatMap(...)`.

Alternatively, you can go all-in with Lodash and do:
```javascript
_.chain(aaData).flatMap(filter.dataParam).uniq().value()
```

alfonsomunozpomer commented on src/ExperimentTable.js:
```diff
@@ -103,33 +131,50 @@ class ExperimentTable extends React.Component {
     const displayedFields = tableHeader.map(header => header.dataParam)
 
-    const dataArray = selectedSearch.trim() ?
+    const experimentTableFilters = tableFilters.map(filter => {
```
Note that `experimentTableFilters` is derived from the prop `aaData`, which means that it won’t change through the lifecycle of the component and doesn’t need to be computed every time the component re-renders. You can move this to the constructor so that it becomes part of the state.

haideriqbal replied:
done

alfonsomunozpomer commented on src/ExperimentTable.js:
```diff
@@ -45,7 +45,7 @@ class ExperimentTable extends React.Component {
       _.sortBy(data, (o) => reverseDateRepresentation(o[propKey])) :
       _.sortBy(data, propKey)
-    return ascendingOrder ? sortedAscendingElements :  sortedAscendingElements.reverse()
+    return ascendingOrder ? sortedAscendingElements : sortedAscendingElements.reverse()
```
Well spotted.

alfonsomunozpomer commented on src/ExperimentTable.js:
```diff
@@ -103,33 +131,50 @@ class ExperimentTable extends React.Component {
+        return selectedDropdownFilters.every(filter => {
+          return filter ? this._isContains(data, filter.value) : true
+        })
+      })
```
There are a few opportunities for simplification here. Firstly, if an arrow function is only a `return ...` statement, you can omit it and the curly braces, and that is implicitly returned. Therefore:
```javascript
const dropdownFilteredExperiments = selectedDropdownFilters.length === 0 ?
  tableHeaderFilteredExperiments : tableHeaderFilteredExperiments.filter(data =>
    selectedDropdownFilters.every(filter =>
        filter ? this._isContains(data, filter.value) : true
    )
)
```

Second, and more importantly, any empty array returns `true` for `.every()`, even if the return value is false (it makes sense if you carefully think about it, honestly!):
```bash
$ node
> [].every(e => false)
true
```

Therefore, if `selectedDropdownFilters` is empty the above block becomes:
```javascript
tableHeaderFilteredExperiments.filter(data =>
    [].every(filter =>
        filter ? this._isContains(data, filter.value) : true
    )
)
```

Further simplifying:
```javascript
tableHeaderFilteredExperiments.filter(data =>
    true
)
```

Which means that when `selectedDropdownFilters` you’ll get the unmodified array `tableHeaderFilteredExperiments`, which means that you don’t need the ternary operator.

haideriqbal replied:
Fixed it!

alfonsomunozpomer commented on src/ExperimentTable.js:
```diff
@@ -60,6 +60,15 @@ class ExperimentTable extends React.Component {
       )
   }
 
+  _isContains(data, value) {
```
This function has a very confusing name. Could you please think of another one? Also add a one sentence description of what it does (or a simple example of when it should either `true` or `false`).

alfonsomunozpomer commented on src/ExperimentTable.js:
```diff
@@ -60,6 +60,15 @@ class ExperimentTable extends React.Component {
 
+  _isContains(data, value) {
+    let contains = false
+    Object.keys(data).some(key => {
```
Room for simplification here:
```javascript
_isContains(data, value) {
  return Object.keys(data).some(key => typeof ...)
}
```

Maybe the logic itself can be simplified, but we can get to that after the “return noise” and throw-away variables are removed.

haideriqbal replied:
Done

alfonsomunozpomer commented on src/ExperimentTable.js:
```diff
@@ -103,33 +131,50 @@ class ExperimentTable extends React.Component {
-        .filter(data => selectedKingdom ? data.kingdom === selectedKingdom : true)
+    const dataArray = selectedSearch.trim() ?
+      _.intersection(selectedSearchFilteredExperiments, dropdownFilteredExperiments) :
+      dropdownFilteredExperiments
```
Think about the logic here: you’re filtering the whole set of results by two different set of criteria, and then doing the interesection of the two result sets. It’s way more efficient and concise filtering by the first set of criteria, and then chain the second `filter` to the remaining elements.

haideriqbal replied:
Fixed

alfonsomunozpomer reviewed (changes requested):
The code does what it should, but we can get away with significant simplifications. Refactor a bit and you’ll get more concise, expressive code.

haideriqbal committed:
- Refactor to optimize code performance

alfonsomunozpomer commented on src/ExperimentTable.js:
```diff
@@ -20,13 +20,19 @@ class ExperimentTable extends React.Component {
+          label: filter.label,
+          options: _.chain(this.props.aaData).flatMap(filter.dataParam).uniq().value()
+        }
+      })
```
Notice that `props` is passed as an argument to `constructor`, so you don’t need to qualify them with `this.`. Replace `this.props` with `props` throughout the constructor.

alfonsomunozpomer commented on src/ExperimentTable.js:
```diff
@@ -103,33 +137,38 @@ class ExperimentTable extends React.Component {
+    const dataArray = _.chain(selectedSearchFilteredExperiments).filter(data =>
+      selectedDropdownFilters.every(filter =>
+        filter ? this.containsValue(data, filter.value) : true
+      )).value()
```
You can do `.filter(...).filter(...).value()`. You can even start with `this.filter(...)`. No need to separate this into several steps. The only thing in this sort of cases is that you should be extra careful and sensible indenting your code, so that it’s easy to see where each step starts and ends. You can add comments in between such as: `// first we filter for the selected dropdowns` or whatever.

Also `dataArray` seems a bit too generic, how about `filteredElements` or `filteredArray`?

haideriqbal replied:
Make sense. Renamed and refactored.

alfonsomunozpomer commented on src/ExperimentTable.js:
```diff
@@ -59,6 +65,15 @@ class ExperimentTable extends React.Component {
+      typeof jsonObject[key] === `object` ?
+        this.containsValue(jsonObject[key], value) : jsonObject[key] === value
+    )
+  }
```
The key point of this function is that it looks for a value *deeply*. Otherwise you could just do `Object.values(jsonObject).includes(value)`.

alfonsomunozpomer reviewed (changes requested):
Please refactor and include a better description for `containsValue`. Very near completion but some minor details missing.

haideriqbal committed:
- Refactor code and add better description for function

alfonsomunozpomer commented on src/ExperimentTable.js:
```diff
@@ -103,33 +137,37 @@ class ExperimentTable extends React.Component {
+        value && value.toString().toLowerCase().includes(selectedSearch))).filter(experiment =>
+      selectedDropdownFilters.every(filter =>
+        filter ? this.containsValue(experiment, filter.value) : true
+      ))
```
How about this? Do you think it reads better?
```javascript
const filteredExperiments =
  this
    // First we filter according to table headers...
    .filter(this.sort(aaData), tableHeader)
    // ... then we further filter according to search box text...
    .filter(experiment =>
      experiment &&
      Object
        .keys(experiment)
        .map(key => displayedFields.includes(key) ? experiment[key] : null)
        .some(value => value && value.toString().toLowerCase().includes(selectedSearch.toLowerCase())))
    // ... and finally we filter according to selected dropdowns
    .filter(experiment =>
      selectedDropdownFilters.every(filter => filter ? this.containsValue(experiment, filter.value) : true))
```

This took me a while and a few tries. Don’t settle for the first thing that comes to mind. Try to be creative while consistent (for example, on your first `filter` block the argument is `data`, while in the second it’s `experiment`, when they’re the same elements). Shuffle things around, try different things.

haideriqbal replied:
This definitely looks better. I have added a little touch of mine. Please review and let me know if it looks fine.

alfonsomunozpomer commented on src/ExperimentTable.js:
```diff
@@ -59,6 +65,15 @@ class ExperimentTable extends React.Component {
           .includes(searchQuery.toLowerCase())
       )
   }
+  //this function searches deeply if value exists in json object e.g.
```
Add an example of a deep search below.

alfonsomunozpomer reviewed (approved):
Some polishing and this is good to go.

haideriqbal committed:
- Refactor code and add new example for containsValue function

alfonsomunozpomer commented on src/ExperimentTable.js:
```diff
@@ -145,16 +158,27 @@ class ExperimentTable extends React.Component {
+              value
+                .toString()
+                .toLowerCase()
+                .includes(selectedSearch)))
```
I just noticed that if you do `value.toLowerCase()` then you should do `selectedSearch.toLowerCase()` too.

haideriqbal committed:
- Apply toLowerCase to selectedSearch

haideriqbal committed:
Merge branch 'master' into feature/168791166-pass-dropdowns-as-props
